x/programs/cmd/simulator/src/lib.rs

```diff
@@ -233,46 +206,47 @@ impl Client {
             panic!("Simulator binary not found, must rebuild simulator");
         }
 
-        Self { path }
-    }
-
-    /// Runs a [Plan] against the simulator and returns vec of result.
-    /// # Errors
-    ///
-    /// Returns an error if the serialization or plan fails.
-    pub fn run_plan(self, plan: Plan) -> Result<Vec<PlanResponse>, ClientError> {
-        let Child {
-            mut stdin,
-            mut stdout,
-            ..
-        } = Command::new(self.path)
+        let Child { stdin, stdout, .. } = Command::new(path)
             .arg("interpreter")
             .arg("--cleanup")
             .arg("--log-level")
             .arg("error")
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .spawn()?;
+            .spawn()
+            .expect("failed to run the simulator in interpreter mode");
 
-        let writer = stdin.as_mut().ok_or(ClientError::StdIo)?;
-        let reader = stdout.as_mut().ok_or(ClientError::StdIo)?;
+        let writer = stdin.expect("stdin has been captured");
+        let reader = stdout.expect("stdout has been captured");
 
         let responses = BufReader::new(reader)
             .lines()
             .map(|line| serde_json::from_str(&line?).map_err(Into::into));
 
-        let mut child = SimulatorChild { writer, responses };
+        let child = SimulatorChild { writer, responses };
+
+        Client { child }
+    }
+}
 
+impl<W, R> Client<W, R>
+where
+    W: Write,
+    R: Iterator<Item = Result<PlanResponse, ClientError>>,
+{
+    /// Runs a [Plan] against the simulator and returns vec of result.
+    /// # Errors
+    ///
+    /// Returns an error if the serialization or plan fails.
+    pub fn run_plan(&mut self, plan: Plan) -> Result<Vec<PlanResponse>, ClientError> {
         plan.steps
             .iter()
-            .map(|step| child.run_step(&plan.caller_key, step))
+            .map(|step| self.child.run_step(plan.caller_key, step))
             .collect()
     }
-}
 
-impl Default for Client {
-    fn default() -> Self {
-        Self::new()
+    pub fn run_step(&mut self, caller_key: &str, step: Step) -> Result<PlanResponse, ClientError> {
+        self.child.run_step(caller_key, &step)
```

Comment:
(context for this comment in the previous comment)

You can move all this stuff to the `Client` (renamed `SimulatorChild`)

Reply:
Thanks! I got a bit of difficulties regarding typing for `PlanResponse` because they are encoded as base64 from Go and we then need to deserialize it but not in the case of a plan, for legacy reasons. This is to keep the data at the state before deserialization.
Does it still makes sense to do the plan removal in a followup PR ?